python/lsst/daf/butler/remote_butler/registry/_query_dimension_records.py

```diff
@@ -0,0 +1,137 @@
+    ``Registry.queryDimensionRecords``.
+    """
+
+    element: DimensionElement
+    dataId: DataId | None
+    dataset_types: list[str]
+    collections: list[str] | None
+    where: str
+    bind: dict[str, Any] | None
+    kwargs: dict[str, DataIdValue]
+
+
+class QueryDriverDimensionRecordQueryResults(LegacyDimensionRecordQueryResults):
+    """Implementation of the legacy ``DimensionRecordQueryResults`` interface
+    using the new query system.
+
+    Parameters
+    ----------
+    query_factory : `QueryFactory`
+        Function that can be called to access the new query system.
+    args : `DimensionRecordsQueryArguments`
+        User-facing arguments forwarded from
+        ``registry.queryDimensionRecords``.
+    """
+
+    def __init__(
+        self,
+        query_factory: QueryFactory,
+        args: DimensionRecordsQueryArguments,
+    ) -> None:
+        self._query_factory = query_factory
+        self._args = args
+        self._limit: int | None = None
+        self._order_by: list[str] = []
+
+    @property
+    def element(self) -> DimensionElement:
+        return self._args.element
+
+    def __iter__(self) -> Iterator[DimensionRecord]:
+        with self._build_query() as result:
+            # We have to eagerly fetch the results to prevent
+            # leaking the resources associated with QueryDriver.
+            records = list(result)
+        yield from records
+
+    def run(self) -> LegacyDimensionRecordQueryResults:
+        return self
+
+    def count(self, *, exact: bool = True, discard: bool = False) -> int:
+        with self._build_query() as result:
+            return result.count(exact=exact, discard=discard)
+
+    def any(self, *, execute: bool = True, exact: bool = True) -> bool:
+        with self._build_query() as result:
+            return result.any(execute=execute, exact=exact)
+
+    def order_by(self, *args: str) -> LegacyDimensionRecordQueryResults:
+        self._order_by.extend(args)
+        return self
+
+    def limit(self, limit: int, offset: int | None = 0) -> LegacyDimensionRecordQueryResults:
+        if offset is not None and offset != 0:
+            raise NotImplementedError("Offset is no longer supported.")
+
+        self._limit = limit
+
+        return self
+
+    def explain_no_results(self, execute: bool = True) -> Iterable[str]:
+        with self._build_query() as result:
+            return result.explain_no_results(execute=execute)
+
+    @contextmanager
+    def _build_query(self) -> Iterator[DimensionRecordQueryResults]:
+        with self._query_factory() as query:
+            a = self._args
+            for dataset_type in a.dataset_types:
+                query = query.join_dataset_search(dataset_type, a.collections)
+            if a.where:
+                query = query.where(a.where, bind=a.bind)
+            if a.dataId:
+                query = query.where(a.dataId)
+            if a.kwargs:
+                query = query.where(a.kwargs)
```

Comment:
We should combine these last two cases into one:
```
if a.dataId:
     kwargs = a.kwargs if a.kwargs else {}  # or use default_factory=dict in dataclass
     query = query.where(a.dataId, **a.kwargs)
```
This lets the values in `kwargs` replace those in `dataId`, which is what users have come to expect from data IDs in kwargs, instead of putting the conflicting constraint into the query and dooming it yield no results.